File: java/liskov-plus-start/src/main/java/no/bouvet/solid/lsp/services/RefundService.java
package no.bouvet.solid.lsp.services;

import java.math.BigDecimal;

import no.bouvet.solid.lsp.domain.PayPalPayment;
import no.bouvet.solid.lsp.domain.Payment;
import no.bouvet.solid.lsp.domain.WorldPayPayment;

public class RefundService {

    public boolean refund(PaymentType paymentServiceType, BigDecimal amount, String transactionId) {
        Payment payment = PaymentFactory.getPayment(paymentServiceType);
        boolean refundSuccess = false;

        if (payment instanceof PayPalPayment) {
            ((PayPalPayment) payment).setUsername("Andras");
            ((PayPalPayment) payment).setPassword("Passw0rd");
        } else if (payment instanceof WorldPayPayment) {
            ((WorldPayPayment) payment).setUsername("Andras");
            ((WorldPayPayment) payment).setPassword("Passw0rd");
            ((WorldPayPayment) payment).setProductId("ABC");
        }

        String serviceResponse = payment.refund(amount, transactionId);

        if (serviceResponse.contains("Auth") || serviceResponse.contains("Success")) {
            refundSuccess = true;
        }
        return refundSuccess;
    }

}

File: dotNet/LSP/LSP/Example2/Services/RefundService.cs
using Solid.LSP.Example2.Domain;

namespace Solid.LSP.Example2.Services
{
    public class RefundService
    {
        public bool Refund(PaymentType paymentServiceType,
                           decimal amount,
                           string transactionId)
        {
            Payment payment = PaymentFactory.GetPayment(paymentServiceType);
            bool refundSuccess = false;

            if (payment is PayPalPayment)
            {
                ((PayPalPayment) payment).Username = "...";
                ((PayPalPayment) payment).Password = "Passw0rd";
            }
            else if (payment is WorldPayPayment)
            {
                ((WorldPayPayment) payment).Username = "...";
                ((WorldPayPayment) payment).Password = "Passw0rd";
                ((WorldPayPayment) payment).ProductId = "ABC";
            }

            string serviceResponse = payment.Refund(amount, transactionId);

            if (serviceResponse.Contains("Auth") || serviceResponse.Contains("Success"))
            {
                refundSuccess = true;
            }

            return refundSuccess;
        }
    }
}
